### PR TITLE
Data Access Layer - Database

### PR DESCRIPTION
**Pull request by jorg3lopez**
# Postgres Data Access Layer 

Adding Postgres database to application

## Issue
#672 

## Checklist

- [x] I have added tests to cover my changes
- [x] I have added logging where useful (with appropriate log level)
- [x] I have added JavaDocs where required

### CODE REVIEW

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public static PostgresDao getInstance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public synchronized void upsertMetadata(
+            String id, String sender, String receiver, String hash, Instant timeReceived)
+            throws SQLException {
+
+        try (Connection conn = connect()) {
```
Comment (basiliskus) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java:46]: You could also include here the PreparedStatement (so it's included in the autoclose scope) like this:
`try (Connection conn = connect();
             PreparedStatement statement = conn.prepareStatement("INSERT INTO metadata VALUES (?, ?, ?, ?, ?)")) {`

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java [hunk 2]
```diff
@@ -24,13 +24,21 @@ public static SendOrderUseCase getInstance() {
...
+
+        var partnerMetadata =
+                new PartnerMetadata(
+                        "uniqueId", "senderName", "receiverName", Instant.now(), "abcd");
         try {
-            savePartnerMetadata(submissionId);
+            logger.logInfo("Trying to save the metadata");
+            partnerMetadataStorage.saveMetadata(partnerMetadata);
+            logger.logInfo("Trying to read the metadata");
+            partnerMetadataStorage.readMetadata(partnerMetadata.uniqueId());
```
Comment (halprin) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java:37]: Looks like the call to `savePartnerMetadata` got clobbered in the merge.

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java [hunk 3]
```diff
@@ -27,8 +27,19 @@ public static SendOrderUseCase getInstance() {
...
+                        "senderName",
+                        "receiverName",
+                        Instant.now(),
+                        "abcd"); // For testing, delete when done
         try {
+            logger.logInfo("Trying to save the metadata");
             savePartnerMetadata(submissionId);
+            logger.logInfo("Trying to read the metadata");
+            partnerMetadataStorage.readMetadata(partnerMetadata.uniqueId());
+
```
Comment (halprin) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java:42]: Just adding back the call to `savePartnerMetadata` isn't correct.  The merge was bad and the real fix is just to remove everything except the start of the `try` block and the call to `savePartnerMetadata`.  E.g....

```suggestion
        try {
            savePartnerMetadata(submissionId);
```

Comment (jorg3lopez) [reply etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java]: Yes, totally agree. We are still using some of that code for testing purposes. We will delete that code before taking it out of draft.

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/EtorDomainRegistration.java [hunk 4]
```diff
@@ -81,14 +85,26 @@ public Map<HttpEndpoint, Function<DomainRequest, DomainResponse>> domainRegistra
 
         if (ApplicationContext.getEnvironment().equalsIgnoreCase("local")) {
             ApplicationContext.register(OrderSender.class, LocalFileOrderSender.getInstance());
-            ApplicationContext.register(
-                    PartnerMetadataStorage.class, FilePartnerMetadataStorage.getInstance());
+
```
Comment (basiliskus) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/EtorDomainRegistration.java:88]: Can we simplify the logic below, and the missing azurestorage instance like this:
```
ApplicationContext.register(AzureClient.class, AzureClient.getInstance());

if (ApplicationContext.getProperty("DB_URL") != null) {
    ApplicationContext.register(SqlDriverManager.class, EtorSqlDriverManager.getInstance());
    ApplicationContext.register(DbDao.class, PostgresDao.getInstance());
    ApplicationContext.register(PartnerMetadataStorage.class, DatabasePartnerMetadataStorage.getInstance());
} else if (ApplicationContext.getEnvironment().equalsIgnoreCase("local")) {
    ApplicationContext.register(PartnerMetadataStorage.class, FilePartnerMetadataStorage.getInstance());
} else {
    ApplicationContext.register(PartnerMetadataStorage.class, AzureStorageAccountPartnerMetadataStorage.getInstance());
}

if (ApplicationContext.getEnvironment().equalsIgnoreCase("local")) {
    ApplicationContext.register(OrderSender.class, LocalFileOrderSender.getInstance());
} else {
    ApplicationContext.register(OrderSender.class, ReportStreamOrderSender.getInstance());
}
```

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java [hunk 5]
```diff
@@ -1,10 +1,12 @@
 package gov.hhs.cdc.trustedintermediary.etor.orders;
 
 import gov.hhs.cdc.trustedintermediary.etor.metadata.EtorMetadataStep;
+import gov.hhs.cdc.trustedintermediary.etor.metadata.PartnerMetadata;
```
Comment (basiliskus) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java]: Are these imports needed?

File: .github/workflows/terraform-deploy_reusable.yml [hunk 6]
```diff
@@ -64,6 +64,18 @@ jobs:
...
+        uses: azure/login@v1
+        with:
+          client-id: ${{ secrets.AZURE_CLIENT_ID }}
+          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
+
+      - name: Run Db migration
+        run: |
+          export PGPASSWORD=$(az account get-access-token --resource-type oss-rdbms --query "[accessToken]" -o tsv)
+          psql "host=$(terraform output -raw database_hostname) port=5432 dbname=postgres user=cdcti-github sslmode=require" -c "CREATE TABLE IF NOT EXISTS metadata (message_id varchar(30), sender varchar(30), receiver varchar(30), hash_of_order varchar(1000), time_received timestamptz); GRANT ALL ON metadata TO azure_pg_admin; ALTER TABLE metadata OWNER TO azure_pg_admin;"
```
Comment (basiliskus) [inline .github/workflows/terraform-deploy_reusable.yml:77]: Would it be possible to have the table creating script in separate sql file and referenced here?

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/DatabasePartnerMetadataStorage.java [hunk 7]
```diff
@@ -17,10 +25,30 @@ public static DatabasePartnerMetadataStorage getInstance() {
     }
 
     @Override
-    public Optional<PartnerMetadata> readMetadata(final String uniqueId) {
-        return Optional.empty();
+    public Optional<PartnerMetadata> readMetadata(final String uniqueId)
+            throws PartnerMetadataException {
+        try {
+            PartnerMetadata data = (PartnerMetadata) dao.fetchMetadata(uniqueId);
+            logger.logInfo(data.receivedSubmissionId());
```
Comment (basiliskus) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/DatabasePartnerMetadataStorage.java:32]: It seems like it's missing a message that specifies what is being logged

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        return conn;
+    }
+
+    public static PostgresDao getInstance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public synchronized void upsertMetadata(
+            String id, String sender, String receiver, String hash, Instant timeReceived)
```
Comment (basiliskus) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java:84]: If the `id` here will always be `receivedSubmissionId`, should we specify it here to make it clear? Same for `fetchMetadata` method

**Review by basiliskus — CHANGES_REQUESTED** [review]

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            throws SQLException {
+
+        try (Connection conn = connect();
+                PreparedStatement statement =
+                        conn.prepareStatement("INSERT INTO metadata VALUES (?, ?, ?, ?, ?)")) {
+            // TODO: Update the below statement to handle on conflict, after we figure out what that
+            // behavior should be
```
[the remark this answers is not in the document]
Comment (jcrichlake) [reply etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java:91]: I would say handle this in a future PR once we have the baseline db in place

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java (hunk 5 above)
Comment (jcrichlake) [reply etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/orders/SendOrderUseCase.java]: Done

File: .github/workflows/terraform-deploy_reusable.yml (hunk 6 above)
Comment (jcrichlake) [reply .github/workflows/terraform-deploy_reusable.yml:77]: Great call out, this is something that will go away in our next chunk of work for the database.  Either Flyway or Liquibase.

**Review by basiliskus — APPROVED** [review]
Great job!

**Pull request merged**